projects/plugins/jetpack-mu-wpcom/src/features/coming-soon/coming-soon.php

```diff
@@ -0,0 +1,236 @@
+
+	include __DIR__ . '/fallback-coming-soon-page.php';
+}
+
+/**
+ * This filter makes sure it's possible to fetch `wpcom_public_coming_soon` option via public-api.wordpress.com.
+ *
+ * @param object $options array Retrieved site options, ready to be returned as API respomnse.
+ *
+ * @return array current value of `wpcom_public_coming_soon`
+ */
+function add_public_coming_soon_to_settings_endpoint_get( $options ) {
+	$wpcom_public_coming_soon            = (int) get_option( 'wpcom_public_coming_soon' );
+	$options['wpcom_public_coming_soon'] = $wpcom_public_coming_soon;
+
+	return $options;
+}
+add_filter( 'site_settings_endpoint_get', __NAMESPACE__ . '\add_public_coming_soon_to_settings_endpoint_get' );
+
+/**
+ * This filter makes sure it's possible to change `wpcom_public_coming_soon` option via public-api.wordpress.com.
+ *
+ * @param object $input Filtered POST input.
+ * @param object $unfiltered_input Raw and unfiltered POST input.
+ *
+ * @return mixed
+ */
+function add_public_coming_soon_to_settings_endpoint_post( $input, $unfiltered_input ) {
+	if ( array_key_exists( 'wpcom_public_coming_soon', $unfiltered_input ) ) {
+		$input['wpcom_public_coming_soon'] = (int) $unfiltered_input['wpcom_public_coming_soon'];
+	}
+	return $input;
+}
+add_filter( 'rest_api_update_site_settings', __NAMESPACE__ . '\add_public_coming_soon_to_settings_endpoint_post', 10, 2 );
+
+/**
+ * Launch the site when the privacy mode changes from public-not-indexed
+ * This can happen due to clicking the launch button from the banner
+ * Or due to manually updating the setting in wp-admin or calypso settings page.
+ *
+ * @param  string $old_value the old value of blog_public.
+ * @param  string $value     the new value of blog_public.
+ * @return bool              whether an update occurred.
+ */
+function disable_coming_soon_on_privacy_change( $old_value, $value ) {
+	if ( 0 !== (int) $old_value || 0 === (int) $value ) {
+		// Do nothing if not moving from public-not-indexed.
+		return false;
+	}
+	update_option( 'wpcom_public_coming_soon', 0 );
+	return true;
+}
+add_action( 'update_option_blog_public', __NAMESPACE__ . '\disable_coming_soon_on_privacy_change', 10, 2 );
+
+/**
+ * Adds the `wpcom_public_coming_soon` option to new sites  if requested by the client.
+ *
+ * @param int    $blog_id    Blog ID.
+ * @param int    $user_id    User ID.
+ * @param string $domain     Site domain.
+ * @param string $path       Site path.
+ * @param int    $network_id Network ID. Only relevant on multi-network installations.
+ * @param array  $meta       Meta data. Used to set initial site options.
+ * @return bool              whether an update occurred.
+ */
+function add_option_to_new_site( $blog_id, $user_id, $domain, $path, $network_id, $meta ) {
+	if ( 0 === $meta['public']
+		&& array_key_exists( 'wpcom_public_coming_soon', $meta['options'] )
+		&& 1 === (int) $meta['options']['wpcom_public_coming_soon'] ) {
```

Comment:
(nitpick) I find it easier to follow the logic when formatted like this:
```suggestion
		&& 1 === (int) $meta['options']['wpcom_public_coming_soon']
	) {
```